### ISSUE TITLE
TinyMCE pwlink does not work in nested repeaters (+Frontend Editing)

### ISSUE THREAD

**Issue by BernhardBaumrock** [processwire/processwire-issues#1706]
### Short description of the issue

I'm developing a page where I want to show a daily schedule of a congress. Something like this:

```
Day 1
9-10: foo
10-11: bar
11-12: baz

Day2
9-10: foo
10-11: bar
11-12: baz
```

So I have multiple days on each congress and multiple slots on each day. This is done with a nested repeater:
![image](https://user-images.githubusercontent.com/8488586/226374355-a9a519e6-879d-47b2-b872-ad1dd8da104a.png)

Now the client asked for the possibility to add PDF uploads that he can link to from within the description text.

I thought that would be easy, just add a PDF field to the repeater item and that's it. Nope... Because the pwlink plugin uses the page id of the congress page (arrow nr. 1) and not the page id of the repeater page!

### Expected behavior

In my case it would be necessary for the pwlink plugin to grab the page id of the repeater page (of the slot) and not of the congress.

### Actual behavior

In the screenshot you see that I want to insert the link in the description of the slot (3) and want to upload the PDF in the field in the slot (4). But PW opens the congress page and then I can't select the uploaded file, which lives on the repeater page and not on the congress page.

### Optional: Suggestion for a possible fix

I've had a look into TinyMCE's `pwlink.js` and I found the function `getPageId()` which I modified like this:

```js
  var selection = editor.selection;
  var node = selection.getNode();

  // added this
  var $repeater = $(node).closest(".InputfieldRepeaterItem");

  var nodeName = node.nodeName.toUpperCase(); // will typically be 'A', 'IMG' or 'P'
  var selectionText = selection.getContent({ format: "text" });
  var selectionHtml = selection.getContent();

  function getPageId() {

    // added this
    if ($repeater) return $repeater.data("page");

    var $in = jQuery("#Inputfield_id");
    var pageId;
    if ($in.length) {
      pageId = $in.val();
    } else {
      pageId = $("#" + editor.id)
        .closest(".Inputfield")
        .attr("data-pid");
    }
    return pageId;
  }
```

This does almost work, but strangely the modal window shows both files (from two different repeater pages):

<img width="728" alt="image" src="https://user-images.githubusercontent.com/8488586/226380388-1163a7df-48f1-4c3c-8666-5c4b148a987f.png">

---

Another thing that I noted is that the current code without my fix works properly when using RockFrontend's ALFRED for frontend editing. There I open the repeater page in a modal and that makes the pwlink plugin grab the correct page id (namely the page id of the repeater page and not the one of the congress page).

So at least the current implementation is not working the same on backend and frontend editing, which is an issue on its own.

I have no proper solution yet but I'd appreciate any help or thoughts on how we could solve this. I can imagine that in some situations using the page id of the congress page rather then the repeater item would be the desired solution. Maybe we need a setting to define which id the plugin should use?

Thanks in advance!

**Comment by ryancramerdesign** [comment]
@BernhardBaumrock When you edit a link in the rich text editor, and that rich text editor happens to be in a repeater, the page ID that it opens ProcessPageEditLink with should be the main page being edited (not the ID of the repeater page). The files dropdown should show all the files associated with that main page, whether the file is in a repeater item, or outside of it. 

So if that's not the behavior you are seeing, it seems like your suggested fix is close to correct. But you are getting the `data-page` of the closest `.InputfieldRepeaterItem`, and what you actually want is the `data-page` of the furthest `.InputfieldRepeater` (not item). Maybe this would do it? (lines that say ADDED)

~~~~~
	function getPageId() {
		var $in = jQuery("#Inputfield_id");
		var pageId;
		if($in.length) {
			pageId = $in.val();
		} else {
			var $repeater = $(node).parents('.InputfieldRepeater').last(); // ADDED 
			if($repeater.length) { // ADDED 
				pageId = $repeater.attr('data-page'); // ADDED
			} else {
				pageId = $("#" + editor.id).closest('.Inputfield').attr('data-pid');
			}
		}
		return pageId;
	}
~~~~~

**Comment by BernhardBaumrock** [comment]
Hi @ryancramerdesign thx for having a look.

What you describe is how it is working on the backend but not on the frontend and it is actually a problem in my scenario:

- I have a page "congress".
- on that page i have a repeater "days".
- every repeater items stands for a day of the congress
- every "day" repeater item shows several "slots"
- every "slot" has fields "title, datetime, description, downloads"

Problem: In the "description" field i can not select the downloads that are added to the "downloads" field in the "slot" repeater item/template.

You are right, i COULD add the downloads field to the "congress" page, which is the page that is shown in the pwlink modal. But this is really less than ideal when using frontend editing, to say the least.

Please see this screenshot how I'm using ALFRED for frontend editing:

<img width="326" alt="image" src="https://user-images.githubusercontent.com/8488586/228843281-1fe1d493-8339-45f4-9077-fde3c67761de.png">

ALFRED is really nothing special, it uses pw-modal to show the page editor for a given page (or repeater item). Please have a look for 1min to the rockfrontend video to get an impression: https://youtu.be/7CoIj--u4ps?t=1955 (the link already starts at the correct spot).

This makes a great editing experience for the user and this is how the modal looks like when he/she clicks on one of the spots:

<img width="867" alt="image" src="https://user-images.githubusercontent.com/8488586/228845408-c320b5b8-19de-45f2-a875-37e5a980d38b.png">

Now if the user adds a link to the description and references a file it looks like this:

<img width="957" alt="image" src="https://user-images.githubusercontent.com/8488586/228846207-53a26229-81c0-4a8e-b35e-c555f925b0bc.png">

That's great! Exactly what I want.

The problem is, that on the Backend, it looks like this:

The nested repeater:
<img width="871" alt="image" src="https://user-images.githubusercontent.com/8488586/228846617-9d8723f8-2220-4ba8-b928-18075df9f99f.png">

And when adding a link:
<img width="808" alt="image" src="https://user-images.githubusercontent.com/8488586/228846791-0e1a7d8b-bd96-47f8-9202-d9724f68c898.png">

So it's the exact same repeater item and the exact same fields but the behaviour is different depending on if we are on the frontend or on the backend.

For my case the behaviour of the frontend is definitely better, but there might also be cases where the behaviour of the backend might be preferable. Ideally it would be configurable! But having two different behaviours is not an option imho :)

I hope it's clear now?

**Comment by ryancramerdesign** [comment]
@BernhardBaumrock The admin behavior you describe is the one it should be using, so we should aim to get consistency with that first. And then if you want it to display files from the current repeater item first, or isolate them or something, we could come back to that. 

Let's say the main page ID is 123 and the repeater item ID is 456.

I don't know how the Alfred works, but the way you describe it, my guess would be that it's opening a modal directly to the page editor for the actual repeater item (id 456)? Rather than the main / `forPageRoot` that owns the repeater (id 123)?. If so, then it's using the `#Inputfield_id` input value of that repeater item (456). So for that case (page editor editing repeater item directly), the correct behavior actually would be to show just the files for the repeater item page (and any nested repeaters within it). Though I'm not sure if that's the context here or not. But if it's doing that, then you probably want to make it use the page editor for the main page (id 123) with the `field=your_repeater_field` populated in the query string instead. That way it would all be in the right context.

**Comment by BernhardBaumrock** [comment]
Hi @ryancramerdesign 

> I don't know how the Alfred works, but the way you describe it, my guess would be that it's opening a modal directly to the page editor for the actual repeater item (id 456)? Rather than the main / forPageRoot that owns the repeater (id 123)?.

Exactly. You can try that in the backend by opening `/page/edit/?id=456` which will look something like this:

<img width="958" alt="image" src="https://user-images.githubusercontent.com/8488586/228862461-c344d241-fa6d-4dea-841d-3c9576ebd801.png">

> If so, then it's using the #Inputfield_id input value of that repeater item (456).

Yep, using 456!

> So for that case (page editor editing repeater item directly), the correct behavior actually would be to show just the files for the repeater item page (and any nested repeaters within it). Though I'm not sure if that's the context here or not.

That's how it behaves, yes. And that's good IMHO. And that's what it is doing on the frontend using ALFRED. Because ALFRED just shows the page editor of 456 in a modal.

The problem is that on the page editor of page 123 the pwlink modal let's me choose files from page 123 whereas when opening the page editor of 456 directly it let's me choose files from 456.

> But if it's doing that, then you probably want to make it use the page editor for the main page (id 123) with the field=your_repeater_field populated in the query string instead. That way it would all be in the right context.

I don't understand what you are trying to say here? Could you please rephrase that or cross-check with what I wrote in the previous quote?

**Comment by ryancramerdesign** [comment]
> That's how it behaves, yes. And that's good IMHO. And that's what it is doing on the frontend using ALFRED. Because ALFRED just shows the page editor of 456 in a modal.

If that's your desired behavior then stick with opening the page editor for id 456. When you do that, 456 is considered the main page, since it's the one being edited. So it should show files for page 456, and files on any nested repeaters within it. Though I'll be honest that repeaters were never intended to be edited this way, directly, but if it suits your particular need then it seems fine to me. 

> The problem is that on the page editor of page 123 the pwlink modal let's me choose files from page 123 whereas when opening the page editor of 456 directly it let's me choose files from 456.

If you are editing page 123 in the page editor then it should show the files for both page 123 and 456 (and any repeaters nested). 

> I don't understand what you are trying to say here? Could you please rephrase that or cross-check with what I wrote in the previous quote?

I'm describing the way the page editor can be used for editing individual fields on a page. This is what the PageFrontEdit module uses. For instance, /processwire/page/edit/?id=123&field=your_repeater_field

**Comment by BernhardBaumrock** [comment]
Hi Ryan,

thank you, I found the issue!

> If you are editing page 123 in the page editor then it should show the files for both page 123 and 456 (and any repeaters nested).

That's where we had different results. I've checked everything on a fresh copy and indeed there it behaves like you describe. In my other installation I'm using RockPageBuilder which is based on InputfieldRepeater but likely messes up the pwlink file selection.

Do you have any pointers for me where I could look for? My InputfieldRockPageBuilder.module.php file uses `class InputfieldRockPageBuilder extends InputfieldRepeater`

I'm closing the issue as it's no issue with PW itself. Sorry for the trouble and thx if you have any quick pointers!

**Comment by ryancramerdesign** [comment]
> Do you have any pointers for me where I could look for? My InputfieldRockPageBuilder.module.php file uses class InputfieldRockPageBuilder extends InputfieldRepeater

FieldtypeRepeater and InputfieldRepeater go along together, so I'd just make sure that there's still a FieldtypeRepeater involved here (or something that extends it). In your class that extends InputfieldRepeater, you'll probably want to add a `$modules->get('InputfieldRepeater');` in its init() method so that the relevant JS/CSS files load. If overriding any methods, make sure you also call the parent::method when needed, such as `renderReady()`, `init()`, if using them.

**Comment by BernhardBaumrock** [comment]
Thx you Ryan I'll check that!

**Comment by MetaTunes** [comment]
>  it's no issue with PW itself. 

Hi @ryancramerdesign, I'm not convinced there isn't a PW issue here. I have a virtually identical issue, except with images, rather than files, as documented [here](https://processwire.com/talk/topic/29488-front-end-edit-of-textarea-images-in-child-page/).
In my case, there are no repeaters - I am using a page table field instead, but otherwise the situation is similar in that the pages in the page table are nested in the display of the main page. So, in the example below, the 'blog-post' page, which just holds the blog text and images, is part of a blog page which also includes all the menus etc.. 

![Screenshot 2024-01-27 201102](https://github.com/processwire/processwire-issues/assets/8653279/36883757-493f-447d-abea-6b986852e611)
If I click "Edit blog-post" then it opens the page editor for the blog-post page and all is fine. However, if I edit the text with the inline front end editor, then the modal with 'images on page' shows the images for the host page, not the blog-post page.

The forum post linked above describes this more fully and also shows a hack to PageFrontEdit as I identified that the problem was that #Inputfield_id was set to the host page, not the page being edited. An alternative to changing PageFrontEdit, would be a modification to pwimage.js, analagous to the pwlink change shown by @BernhardBaumrock , namely:
```	
      function buildQueryString() {
		var imgWidth, imgHeight, imgDescription, imgLink, imgClass, nodeClass, hidpi, file = '';
		var $editor = $('#' + editor.id);
		var $inputfield = $editor.closest('.Inputfield');
		var $node = $(node);
		// MDE added line below
		var $editPage = $node.closest(".pw-edit");
		//MDE added condition below and placed the original statements in the 'else' part
		if($editPage) {
			pageId = $editPage.attr('data-page');
		} else {
			var $in = $("#Inputfield_id");
			var pageId = $in.length ? $in.val() : $inputfield.attr('data-pid');
		}
		//MDE end of changes
		var editPageId = pageId;
```

Bottom line is that I'm not sure where to go wiith this next as the resolution above seems to only relate to repeaters.

**Comment by MetaTunes** [comment]
Actiually, that code fix above is not wuite right as it doesn't always work in the back end. This seems better:
```
		// MDE added line below
		var $editPage = $node.closest(".pw-edit");
		//MDE added condition below and placed the original statements in the 'if(!pageId)' part
		if($editPage) {
			pageId = $editPage.attr('data-page');
		}
		if(!pageId) {
			var $in = $("#Inputfield_id");
			var pageId = $in.length ? $in.val() : $inputfield.attr('data-pid');
		}
		//MDE end of changes
```
